File: src/HomeManagementModule.ts
import { NativeModule, requireNativeModule } from 'expo';

import {
    HomeBean,
    WeatherSketch,
    WeatherDetail,
    WeatherUnits,
    HomeManagementModuleEvents,
    MemberBean,
    AddMemberParams,
    UpdateMemberParams,
    InvitationRecord,
} from './HomeManagement.types';

declare class HomeManagementModule extends NativeModule<HomeManagementModuleEvents> {
    // ─── Home Manager ───────────────────────────────────────────────────

    /**
     * Create a new home.
     * @param name - Home name (max 25 chars).
     * @param lon - Longitude (0 if not needed).
     * @param lat - Latitude (0 if not needed).
     * @param geoName - Address/location name.
     * @param rooms - List of room names.
     * @returns The homeId of the created home.
     */
    createHome(
        name: string,
        lon: number,
        lat: number,
        geoName: string,
        rooms: string[]
    ): Promise<number>;

    /**
     * Query the list of homes for the current user.
     */
    queryHomeList(): Promise<HomeBean[]>;

    // ─── Home Instance ──────────────────────────────────────────────────

    /**
     * Get the full details of a home (devices, groups, rooms, etc.).
     * @param homeId - The home ID.
     */
    getHomeDetail(homeId: number): Promise<HomeBean>;

    /**
     * Get cached/offline home details.
     * @param homeId - The home ID.
     */
    getHomeLocalCache(homeId: number): Promise<HomeBean>;

    /**
     * Update home information.
     * @param homeId - The home ID.
     * @param name - New home name.
     * @param lon - Longitude.
     * @param lat - Latitude.
     * @param geoName - Address/location name.
     * @param rooms - List of room names.
     */
    updateHome(
        homeId: number,
        name: string,
        lon: number,
        lat: number,
        geoName: string,
        rooms: string[]
    ): Promise<void>;

    /**
     * Dismiss (delete) a home.
     * @param homeId - The home ID.
     */
    dismissHome(homeId: number): Promise<void>;

    // ─── Weather ────────────────────────────────────────────────────────

    /**
     * Get weather overview for the home's location.
     * @param homeId - The home ID.
     */
    getHomeWeatherSketch(homeId: number): Promise<WeatherSketch>;

    /**
     * Get detailed weather for the home's location.
     * @param homeId - The home ID.
     * @param limit - Max number of weather metrics to return.
     * @param units - Unit preferences (tempUnit, pressureUnit, windspeedUnit).
     */
    getHomeWeatherDetail(
        homeId: number,
        limit: number,
        units: WeatherUnits
    ): Promise<WeatherDetail[]>;

    // ─── Sorting ────────────────────────────────────────────────────────

    /**
     * Sort devices and groups in a home.
     * @param homeId - The home ID.
     * @param orderList - Array of { bizId: string, bizType: string }.
     *   bizType: "6" for device, "5" for group.
     */
    sortDeviceOrGroup(
        homeId: number,
        orderList: Array<{ bizId: string; bizType: string }>
    ): Promise<void>;

    // ─── Room Management ────────────────────────────────────────────────

    /**
     * Add a room to a home.
     * @param homeId - The home ID.
     * @param name - The room name.
     */
    addRoom(homeId: number, name: string): Promise<number>;

    /**
     * Remove a room from a home.
     * @param homeId - The home ID.
     * @param roomId - The room ID.
     */
    removeRoom(homeId: number, roomId: number): Promise<void>;

    /**
     * Sort rooms in a home.
     * @param homeId - The home ID.
     * @param roomIds - Ordered list of room IDs.
     */
    sortRoomList(homeId: number, roomIds: number[]): Promise<void>;

    /**
     * Rename a room.
     * @param roomId - The room ID.
     * @param name - The new room name.
     */
    updateRoomName(roomId: number, name: string): Promise<void>;

    /**
     * Add a device to a room.
     * @param roomId - The room ID.
     * @param devId - The device ID.
     */
    addDeviceToRoom(roomId: number, devId: string): Promise<void>;

    /**
     * Remove a device from a room.
     * @param roomId - The room ID.
     * @param devId - The device ID.
     */
    removeDeviceFromRoom(roomId: number, devId: string): Promise<void>;

    /**
     * Add a group to a room.
     * @param roomId - The room ID.
     * @param groupId - The group ID.
     */
    addGroupToRoom(roomId: number, groupId: string): Promise<void>;

    /**
     * Remove a group from a room.
     * @param roomId - The room ID.
     * @param groupId - The group ID.
     */
    removeGroupFromRoom(roomId: number, groupId: string): Promise<void>;

    /**
     * Batch update room-device/group mappings.
     * @param roomId - The room ID.
     * @param deviceGroupIds - List of device IDs or group IDs.
     */
    saveBatchRoomRelation(roomId: number, deviceGroupIds: string[]): Promise<void>;

    // ─── Member Management ────────────────────────────────────────────

    /**
     * Add a member to a home.
     * @param params - Member info (homeId, nickName, account, countryCode, role, autoAccept).
     */
    addMember(params: AddMemberParams): Promise<MemberBean>;

    /**
     * Remove a member from a home.
     * @param memberId - The member ID.
     */
    removeMember(memberId: number): Promise<void>;

    /**
     * Query the list of members in a home.
     * @param homeId - The home ID.
     */
    queryMemberList(homeId: number): Promise<MemberBean[]>;

    /**
     * Update member information.
     * @param params - Updated member info (memberId, nickName, role, admin).
     */
    updateMember(params: UpdateMemberParams): Promise<void>;

    /**
     * Get an invitation code/message for a home.
     * @param homeId - The home ID.
     */
    getInvitationCode(homeId: number): Promise<string>;

    /**
     * Join a home using an invitation code.
     * @param invitationCode - The invitation code.
     */
    joinHomeWithInvitationCode(invitationCode: string): Promise<void>;

    /**
     * Cancel/revoke an invitation.
     * @param invitationId - The invitation ID.
     */
    cancelInvitation(invitationId: number): Promise<void>;

    /**
     * Accept or decline an invitation to join a home.
     * @param homeId - The home ID.
     * @param accept - true to accept, false to decline.
     */
    processInvitation(homeId: number, accept: boolean): Promise<void>;

    /**
     * Query invitation records for a home.
     * @param homeId - The home ID.
     */
    getInvitationList(homeId: number): Promise<InvitationRecord[]>;

    /**
     * Update invitee information (name, role).
     * @param invitationId - The invitation ID.
     * @param name - The new name.
     * @param role - The new role.
     */
    updateInvitedMember(invitationId: number, name: string, role: number): Promise<void>;
}

export default requireNativeModule<HomeManagementModule>('ExpoTuyaHomeManagement');

File: ios/HomeManagementModule.swift
import ExpoModulesCore
import ThingSmartDeviceKit

public class HomeManagementModule: Module {
  private var homeManager: ThingSmartHomeManager?
  private var homeInstances: [Int64: ThingSmartHome] = [:]

  public func definition() -> ModuleDefinition {
    Name("ExpoTuyaHomeManagement")

    Events(
      "onHomeAdded",
      "onHomeRemoved",
      "onHomeInfoChanged",
      "onHomeInvite",
      "onServerConnectSuccess",
      "onDeviceAdded",
      "onDeviceRemoved",
      "onGroupAdded",
      "onGroupRemoved"
    )

    // MARK: - Create a home

    AsyncFunction("createHome") { (name: String, lon: Double, lat: Double, geoName: String, rooms: [String], promise: Promise) in
      self.getHomeManager().addHome(
        withName: name,
        geoName: geoName,
        rooms: rooms,
        latitude: lat,
        longitude: lon,
        success: { homeId in
          promise.resolve(homeId)
        },
        failure: { error in
          promise.reject(error ?? self.makeError("Failed to create home"))
        }
      )
    }

    // MARK: - Query home list

    AsyncFunction("queryHomeList") { (promise: Promise) in
      self.getHomeManager().getHomeList(success: { homes in
        let result = homes?.map { self.homeModelToDict($0) } ?? []
        promise.resolve(result)
      }, failure: { error in
        promise.reject(error ?? self.makeError("Failed to query home list"))
      })
    }

    // MARK: - Get home detail

    AsyncFunction("getHomeDetail") { (homeId: Int64, promise: Promise) in
      let home = self.getHomeInstance(homeId)
      home.getHomeData(success: { homeModel in
        promise.resolve(self.homeModelToDict(homeModel))
      }, failure: { error in
        promise.reject(error ?? self.makeError("Failed to get home detail"))
      })
    }

    // MARK: - Get home local cache

    AsyncFunction("getHomeLocalCache") { (homeId: Int64, promise: Promise) in
      let home = self.getHomeInstance(homeId)
      home.getHomeLocalCache(success: { homeModel in
        promise.resolve(self.homeModelToDict(homeModel))
      }, failure: { error in
        promise.reject(error ?? self.makeError("Failed to get home cache"))
      })
    }

    // MARK: - Update home

    AsyncFunction("updateHome") { (homeId: Int64, name: String, lon: Double, lat: Double, geoName: String, rooms: [String], promise: Promise) in
      let home = self.getHomeInstance(homeId)
      home.updateHomeInfo(
        withName: name,
        geoName: geoName,
        latitude: lat,
        longitude: lon,
        success: {
          promise.resolve(nil)
        },
        failure: { error in
          promise.reject(error ?? self.makeError("Failed to update home"))
        }
      )
    }

    // MARK: - Dismiss home

    AsyncFunction("dismissHome") { (homeId: Int64, promise: Promise) in
      let home = self.getHomeInstance(homeId)
      home.dismissHome(success: {
        self.homeInstances.removeValue(forKey: homeId)
        promise.resolve(nil)
      }, failure: { error in
        promise.reject(error ?? self.makeError("Failed to dismiss home"))
      })
    }

    // MARK: - Weather sketch

    AsyncFunction("getHomeWeatherSketch") { (homeId: Int64, promise: Promise) in
      let home = self.getHomeInstance(homeId)
      home.getHomeWeatherSketch(success: { sketch in
        promise.resolve([
          "condition": sketch?.condition ?? "",
          "temp": sketch?.temp ?? "",
          "iconUrl": sketch?.iconUrl ?? "",
          "inIconUrl": sketch?.inIconUrl ?? "",
        ])
      }, failure: { error in
        promise.reject(error ?? self.makeError("Failed to get weather sketch"))
      })
    }

    // MARK: - Weather detail

    AsyncFunction("getHomeWeatherDetail") { (homeId: Int64, limit: Int, units: [String: Any], promise: Promise) in
      let home = self.getHomeInstance(homeId)
      let optionModel = ThingSmartWeatherOptionModel()
      if let tempUnit = units["tempUnit"] as? Int {
        optionModel.temperatureUnit = ThingSmartTemperatureUnit(rawValue: tempUnit) ?? .celsius
      }
      if let pressureUnit = units["pressureUnit"] as? Int {
        optionModel.pressureUnit = ThingSmartPressureUnit(rawValue: pressureUnit) ?? .hPa
      }
      if let windspeedUnit = units["windspeedUnit"] as? Int {
        optionModel.windspeedUnit = ThingSmartWindspeedUnit(rawValue: windspeedUnit) ?? .mPerSec
      }
      optionModel.limit = limit

      home.getHomeWeatherDetail(with: optionModel, success: { models in
        let result = models?.map { model -> [String: Any] in
          return [
            "icon": model.icon ?? "",
            "name": model.name ?? "",
            "unit": model.unit ?? "",
            "value": model.value ?? "",
          ]
        } ?? []
        promise.resolve(result)
      }, failure: { error in
        promise.reject(error ?? self.makeError("Failed to get weather detail"))
      })
    }

    // MARK: - Sort devices/groups

    AsyncFunction("sortDeviceOrGroup") { (homeId: Int64, orderList: [[String: String]], promise: Promise) in
      let home = self.getHomeInstance(homeId)
      let nsOrderList = orderList.map { item -> NSDictionary in
        return item as NSDictionary
      }
      home.sortDeviceOrGroup(withOrderList: nsOrderList, success: {
        promise.resolve(nil)
      }, failure: { error in
        promise.reject(error ?? self.makeError("Failed to sort devices/groups"))
      })
    }

    // MARK: - Add room

    AsyncFunction("addRoom") { (homeId: Int64, name: String, promise: Promise) in
      let home = self.getHomeInstance(homeId)
      home.addHomeRoom(withName: name, success: {
        // iOS doesn't return roomId directly from this callback, resolve nil
        promise.resolve(0)
      }, failure: { error in
        promise.reject(error ?? self.makeError("Failed to add room"))
      })
    }

    // MARK: - Remove room

    AsyncFunction("removeRoom") { (homeId: Int64, roomId: Int64, promise: Promise) in
      let home = self.getHomeInstance(homeId)
      home.removeHomeRoom(withRoomId: roomId, success: {
        promise.resolve(nil)
      }, failure: { error in
        promise.reject(error ?? self.makeError("Failed to remove room"))
      })
    }

    // MARK: - Sort rooms

    AsyncFunction("sortRoomList") { (homeId: Int64, roomIds: [Int64], promise: Promise) in
      let home = self.getHomeInstance(homeId)
      // Build ThingSmartRoomModel list from roomIds using home's roomList
      let roomModels = roomIds.compactMap { rid -> ThingSmartRoomModel? in
        return home.roomList?.first(where: { $0.roomId == rid })
      }
      home.sortRoomList(roomModels, success: {
        promise.resolve(nil)
      }, failure: { error in
        promise.reject(error ?? self.makeError("Failed to sort rooms"))
      })
    }

    // MARK: - Update room name

    AsyncFunction("updateRoomName") { (roomId: Int64, name: String, promise: Promise) in
      let room = ThingSmartRoom(roomId: roomId)
      room?.updateRoomName(name, success: {
        promise.resolve(nil)
      }, failure: { error in
        promise.reject(error ?? self.makeError("Failed to update room name"))
      })
    }

    // MARK: - Add device to room

    AsyncFunction("addDeviceToRoom") { (roomId: Int64, devId: String, promise: Promise) in
      let room = ThingSmartRoom(roomId: roomId)
      room?.addDevice(withDeviceId: devId, success: {
        promise.resolve(nil)
      }, failure: { error in
        promise.reject(error ?? self.makeError("Failed to add device to room"))
      })
    }

    // MARK: - Remove device from room

    AsyncFunction("removeDeviceFromRoom") { (roomId: Int64, devId: String, promise: Promise) in
      let room = ThingSmartRoom(roomId: roomId)
      room?.removeDevice(withDeviceId: devId, success: {
        promise.resolve(nil)
      }, failure: { error in
        promise.reject(error ?? self.makeError("Failed to remove device from room"))
      })
    }

    // MARK: - Add group to room

    AsyncFunction("addGroupToRoom") { (roomId: Int64, groupId: String, promise: Promise) in
      let room = ThingSmartRoom(roomId: roomId)
      room?.addGroup(withGroupId: groupId, success: {
        promise.resolve(nil)
      }, failure: { error in
        promise.reject(error ?? self.makeError("Failed to add group to room"))
      })
    }

    // MARK: - Remove group from room

    AsyncFunction("removeGroupFromRoom") { (roomId: Int64, groupId: String, promise: Promise) in
      let room = ThingSmartRoom(roomId: roomId)
      room?.removeGroup(withGroupId: groupId, success: {
        promise.resolve(nil)
      }, failure: { error in
        promise.reject(error ?? self.makeError("Failed to remove group from room"))
      })
    }

    // MARK: - Batch room relation

    AsyncFunction("saveBatchRoomRelation") { (roomId: Int64, deviceGroupIds: [String], promise: Promise) in
      let room = ThingSmartRoom(roomId: roomId)
      room?.saveBatchRoomRelation(withDeviceGroupList: deviceGroupIds, success: {
        promise.resolve(nil)
      }, failure: { error in
        promise.reject(error ?? self.makeError("Failed to save batch room relation"))
      })
    }

    // MARK: - Add member

    AsyncFunction("addMember") { (params: [String: Any], promise: Promise) in
      let homeId = params["homeId"] as? Int64 ?? 0
      let home = self.getHomeInstance(homeId)
      let requestModel = ThingSmartHomeAddMemberRequestModel()
      requestModel.name = params["nickName"] as? String ?? ""
      requestModel.account = params["account"] as? String ?? ""
      requestModel.countryCode = params["countryCode"] as? String ?? ""
      requestModel.role = ThingHomeRoleType(rawValue: params["role"] as? UInt ?? 0) ?? .member
      requestModel.autoAccept = params["autoAccept"] as? Bool ?? true

      home.addHomeMember(withAddMemeberRequestModel: requestModel, success: { dict in
        promise.resolve(dict ?? [:])
      }, failure: { error in
        promise.reject(error ?? self.makeError("Failed to add member"))
      })
    }

    // MARK: - Remove member

    AsyncFunction("removeMember") { (memberId: Int64, promise: Promise) in
      let member = ThingSmartHomeMember()
      member.removeHomeMember(withMemberId: memberId, success: {
        promise.resolve(nil)
      }, failure: { error in
        promise.reject(error ?? self.makeError("Failed to remove member"))
      })
    }

    // MARK: - Query member list

    AsyncFunction("queryMemberList") { (homeId: Int64, promise: Promise) in
      let home = self.getHomeInstance(homeId)
      home.getHomeMemberList(success: { members in
        let result = members?.map { m -> [String: Any] in
          return [
            "homeId": homeId,
            "memberId": m.memberId,
            "nickName": m.nickName ?? "",
            "account": m.account ?? "",
            "role": m.role.rawValue,
            "admin": m.role == .admin || m.role == .owner,
            "headPic": m.headPic ?? "",
            "memberStatus": m.dealStatus.rawValue,
          ]
        } ?? []
        promise.resolve(result)
      }, failure: { error in
        promise.reject(error ?? self.makeError("Failed to query member list"))
      })
    }

    // MARK: - Update member

    AsyncFunction("updateMember") { (params: [String: Any], promise: Promise) in
      let member = ThingSmartHomeMember()
      let requestModel = ThingSmartHomeMemberRequestModel()
      requestModel.memberId = params["memberId"] as? Int64 ?? 0
      requestModel.name = params["nickName"] as? String ?? ""
      requestModel.role = ThingHomeRoleType(rawValue: params["role"] as? UInt ?? 0) ?? .member

      member.updateHomeMemberInfo(withMemberRequestModel: requestModel, success: {
        promise.resolve(nil)
      }, failure: { error in
        promise.reject(error ?? self.makeError("Failed to update member"))
      })
    }

    // MARK: - Get invitation code

    AsyncFunction("getInvitationCode") { (homeId: Int64, promise: Promise) in
      let invitation = ThingSmartHomeInvitation()
      let requestModel = ThingSmartHomeInvitationCreateRequestModel()
      requestModel.homeID = homeId
      requestModel.needMsgContent = true
      invitation.createInvitation(with: requestModel, success: { resultModel in
        promise.resolve(resultModel?.invitationCode ?? "")
      }, failure: { error in
        promise.reject(error ?? self.makeError("Failed to get invitation code"))
      })
    }

    // MARK: - Join home with invitation code

    AsyncFunction("joinHomeWithInvitationCode") { (invitationCode: String, promise: Promise) in
      let invitation = ThingSmartHomeInvitation()
      invitation.joinHome(withInvitationCode: invitationCode, success: { result in
        promise.resolve(nil)
      }, failure: { error in
        promise.reject(error ?? self.makeError("Failed to join home"))
      })
    }

    // MARK: - Cancel invitation

    AsyncFunction("cancelInvitation") { (invitationId: Int64, promise: Promise) in
      let invitation = ThingSmartHomeInvitation()
      invitation.cancelInvitation(withInvitationID: NSNumber(value: invitationId), success: { result in
        promise.resolve(nil)
      }, failure: { error in
        promise.reject(error ?? self.makeError("Failed to cancel invitation"))
      })
    }

    // MARK: - Process invitation (accept/decline)

    AsyncFunction("processInvitation") { (homeId: Int64, accept: Bool, promise: Promise) in
      let home = self.getHomeInstance(homeId)
      home.joinFamily(withAccept: accept, success: { result in
        promise.resolve(nil)
      }, failure: { error in
        promise.reject(error ?? self.makeError("Failed to process invitation"))
      })
    }

    // MARK: - Get invitation list

    AsyncFunction("getInvitationList") { (homeId: Int64, promise: Promise) in
      let invitation = ThingSmartHomeInvitation()
      invitation.fetchInvitationRecordList(withHomeID: homeId, success: { records in
        let result = records?.map { r -> [String: Any] in
          return [
            "invitationId": r.invitationID ?? 0,
            "invitationCode": r.invitationCode ?? "",
            "name": r.name ?? "",
            "role": r.role,
            "dealStatus": r.dealStatus,
          ]
        } ?? []
        promise.resolve(result)
      }, failure: { error in
        promise.reject(error ?? self.makeError("Failed to get invitation list"))
      })
    }

    // MARK: - Update invited member

    AsyncFunction("updateInvitedMember") { (invitationId: Int64, name: String, role: Int, promise: Promise) in
      let invitation = ThingSmartHomeInvitation()
      let requestModel = ThingSmartHomeInvitationInfoRequestModel()
      requestModel.invitationID = NSNumber(value: invitationId)
      requestModel.name = name
      requestModel.role = ThingHomeRoleType(rawValue: UInt(role)) ?? .member

      invitation.updateInvitationInfo(with: requestModel, success: { result in
        promise.resolve(nil)
      }, failure: { error in
        promise.reject(error ?? self.makeError("Failed to update invited member"))
      })
    }

    // MARK: - Lifecycle

    OnStartObserving {
      self.getHomeManager().delegate = self
    }

    OnStopObserving {
      self.homeManager?.delegate = nil
    }
  }

  // MARK: - Helpers

  private func getHomeManager() -> ThingSmartHomeManager {
    if homeManager == nil {
      homeManager = ThingSmartHomeManager()
    }
    return homeManager!
  }

  private func getHomeInstance(_ homeId: Int64) -> ThingSmartHome {
    if let existing = homeInstances[homeId] {
      return existing
    }
    let home = ThingSmartHome(homeId: homeId)!
    homeInstances[homeId] = home
    return home
  }

  private func homeModelToDict(_ model: ThingSmartHomeModel?) -> [String: Any] {
    guard let m = model else { return [:] }
    return [
      "homeId": m.homeId,
      "name": m.name ?? "",
      "geoName": m.geoName ?? "",
      "lon": m.longitude,
      "lat": m.latitude,
      "admin": m.admin,
      "homeStatus": m.homeStatus,
      "role": m.role,
    ]
  }

  private func makeError(_ message: String) -> NSError {
    return NSError(domain: "ExpoTuyaHomeManagement", code: -1, userInfo: [NSLocalizedDescriptionKey: message])
  }
}

// MARK: - ThingSmartHomeManagerDelegate

extension HomeManagementModule: ThingSmartHomeManagerDelegate {
  public func homeManager(_ manager: ThingSmartHomeManager, didAddHome home: ThingSmartHomeModel) {
    sendEvent("onHomeAdded", ["homeId": home.homeId])
  }

  public func homeManager(_ manager: ThingSmartHomeManager, didRemoveHome homeId: Int64) {
    homeInstances.removeValue(forKey: homeId)
    sendEvent("onHomeRemoved", ["homeId": homeId])
  }

  public func serviceConnectedSuccess() {
    sendEvent("onServerConnectSuccess")
  }
}
